package/MDAnalysis/lib/formats/xdrlib.pyx

```diff
@@ -0,0 +1,653 @@
+        IOError
+            Couldn't read the file
+        """
+        if self.is_open:
+            self.close()
+        cdef int return_code = 1
+        self.fname = fname
+        self.n_atoms = 0
+        self.reached_eof = False
+        self.current_frame = 0
+
+        if mode == 'r':
+            opening_mode = b'r'
+        elif mode == 'w':
+            opening_mode = b'w'
+        else:
+            raise ValueError('mode must be one of "r" or "w", you '
+                             'supplied {}'.format(mode))
+        self.mode = mode
+
+        if self.mode == 'r':
+            if not exists(self.fname):
+                raise IOError('File does not exists: {}'.format(self.fname))
+
+            return_code, self.n_atoms = self._calc_natoms(fname);
+
+            if return_code != 0:
+                raise IOError('XDRLIB read error: {}'.format(
+                    error_message[return_code]))
+            if self.n_atoms <= 0:
+                raise IOError('Couldn\'t read number of atoms: {}'.format(
+                    fname))
+
+        self.xfp = xdrfile_open(fname, opening_mode)
+        if self.xfp is NULL:
+            raise IOError('error opening xtf file: {}'.format(self.fname))
+        self.is_open = True
+
+    def __enter__(self):
+        """Support context manager"""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Support context manager"""
+        self.close()
+        # always propagate exceptions forward
+        return False
+
+    def __iter__(self):
+        self.close()
+        self.open(self.fname, self.mode)
+        return self
+
+    def __next__(self):
+        if self.reached_eof:
+            raise StopIteration
+        return self.read()
+
+    def __len__(self):
+        if not self.is_open:
+            raise RuntimeError('No file currently opened')
+        return self.offsets.size
+
+    def seek(self, frame):
+        """Seek to Frame.
+
+        Please note if that this function will generate internal file offsets if
+        they haven't been set before. For large file this means the first seek
+        can be very slow. Later seeks will be very fast
+
+        Parameters
+        ----------
+        frame: int
+            seek the file to given frame
+
+        Raises
+        ------
+        RuntimeError
+            If you seek for more frames then are available
+        """
+        cdef int offset
+        if frame == 0:
+            offset = 0
+        else:
+            if frame >= self.offsets.size:
+                raise RuntimeError('Trying to seek over max number of frames')
+            offset = self.offsets[frame]
+        self.reached_eof = False
+        ok = xdr_seek(self.xfp, offset, SEEK_SET)
+        if ok != 0:
+            raise RuntimeError("XDR seek failed, error-message={}".format(
+                error_message[ok]))
+        self.current_frame = frame
+
+    @property
+    def offsets(self):
```

Comment:
Plus I can now grab where the condition for the caching of the value is checked. Looking in my editor for the variable `has_offsets` will show me everywhere I set or reset it. With the cached decorator it is hard to look for then I set `_cache[key]` in the code without knowing that this belongs to the `cached` decorator machinery.